daemon/api_snapshots.go

```diff
@@ -137,3 +142,49 @@ func changeSnapshots(c *Command, r *http.Request, user *auth.UserState) Response
 
 	return AsyncResponse(nil, &Meta{Change: chg.ID()})
 }
+
+type snapshotExportResponse struct {
+	SetID      uint64
+	ExportSize uint64
+}
+
+func (s snapshotExportResponse) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.Header().Add("Content-Length", strconv.FormatUint(s.ExportSize, 10))
+
+	snapshotExport(context.TODO(), uint64(s.SetID), w)
+}
+
+type countingOnlyWriter struct {
+	total uint64
+}
+
+func (w *countingOnlyWriter) Write(p []byte) (n int, err error) {
+	w.total += uint64(len(p))
+	return len(p), nil
+}
+
+func getSnapshotExport(c *Command, r *http.Request, user *auth.UserState) Response {
+	vars := muxVars(r)
+	sid := vars["id"]
+	setID, err := strconv.ParseUint(sid, 10, 64)
+	if err != nil {
+		return BadRequest("'id' must be a positive base 10 number; got %q", sid)
+	}
+
+	// Export into /dev/null once to get the size of the tar so that
+	// we can set the Content-Length in the response
+	//
+	// XXX: too naive? i.e. calling snapshotExport() twice will lead to
```

Comment:
I think actually that this is okay, since we are not actually computing the snapshot here, we are just packaging it up, and as you say the only things that change is the metadata on the snapshot, which should be of fixed length and thus will result in the same Content-Length header value. 

I would feel more comfortable about this if we had a spread test or a unit test which verified just that the same snapshot exported at two different times has the same size.

Reply:
I added a unit test for this now - and it is indeed the case that a clock jump beyond the year 2242 will break this assumption because go will pick a different tar header in this case apparently.

Reply:
/me adds a reminder for 222 years from now to fix this bug